hw3/minchuncho/matrix.cpp

```diff
@@ -0,0 +1,207 @@
+Matrix::Matrix(Matrix const& mat)
+{
+    this->row_ = mat.row_;
+    this->col_ = mat.col_;
+    this->vec_ = mat.vec_;
+}
+
+// copy assignment
+Matrix& Matrix::operator=(Matrix const& mat)
+{
+    if(this == &mat) return *this;
+    
+    this->row_ = mat.row_;
+    this->col_ = mat.col_;
+    this->vec_ = mat.vec_;
+    
+    return *this;
+}
+
+// move constructor
+Matrix::Matrix(Matrix && mat)
+{
+    std::swap(this->row_, mat.row_);
+    std::swap(this->col_, mat.col_);
+    this->vec_ = std::move(mat.vec_);
+}
+
+// move assignment
+Matrix& Matrix::operator=(Matrix && mat)
+{
+    if(this == &mat) return *this;
+    
+    std::swap(this->row_, mat.row_);
+    std::swap(this->col_, mat.col_);
+    this->vec_ = std::move(mat.vec_);
+    
+    return *this;
+}
+
+Matrix& Matrix::operator=(std::vector<double> const& vec)
+{
+    this->vec_ = vec;
+    return *this;
+}
+
+double& Matrix::operator()(size_t const& i, size_t const& j)
+{
+    return this->vec_[i*this->col_ + j];
+}
+
+double Matrix::operator()(size_t const& i, size_t const& j) const
+{
+    return this->vec_[i*this->col_ + j];
+}
+
+bool Matrix::operator==(Matrix const& mat) const
+{
+    if(this->col_ != mat.col_ || this->row_ != mat.row_) return false;
+    
+    for(size_t i=0; i<this->size(); i++){
+        // a floating-point precision issue
+        // if(this->vec_[i] != mat.vec_[i]) return false;
+        if(fabs(this->vec_[i]-mat.vec_[i]) > 0.000001f) return false;
+    }
+    
+    return true;
+}
+
+bool Matrix::bound_check(size_t const& i, size_t const& j)
+{
+    return i<this->row_ && i>=0 && j<this->col_ && j>=0;
+}
+
+double Matrix::get_element(std::pair<size_t, size_t> index)
+{
+    return (*this)(index.first, index.second);
+}
+
+void Matrix::set_element(std::pair<size_t, size_t> index, const double& item)
+{
+    (*this)(index.first, index.second) = item;
+}
+
+std::string Matrix::get_matrix_str()
+{
+    std::string s;
+    for(size_t i=0; i<row_; i++){
+        for(size_t j=0; j<col_; j++){
+            s += std::to_string((*this)(i, j)) + ' ';
+        }
+        s += '\n';
+    }
+    return s;
+}
+
+std::ostream& operator<<(std::ostream& os, Matrix mat)
+{
+    for(size_t i=0; i<mat.row_; i+=mat.col_){
+        for(size_t j=0; j<mat.col_; j++){
+            std::cout << mat(i, j) << " ";
+        }
+        std::cout << std::endl;
+    }
+    
+    return os;
+}
+
+Matrix multiply_naive(Matrix const& mat1, Matrix const& mat2)
+{
+    if(mat1.col() != mat2.row()){
+        throw std::out_of_range("mat1 col is different from mat2 row");
+    }
+
+    Matrix ret(mat1.row(), mat2.col());
+
+    for(size_t i=0; i<mat1.row(); i++){
+        for(size_t j=0; j<mat2.col(); j++){
+            double v = 0;
+            for(size_t k=0; k<mat1.col(); k++){
+                v += mat1(i, k)*mat2(k, j);
+            }
+            ret(i,j) = v;
+        }
+    }
+
+    return ret;
+}
+
+Matrix multiply_tile(Matrix const& mat1, Matrix const& mat2, size_t tsize)
```

Comment:
Hi Professor Chen @yungyuc , I have a question about the function multiply_tile()

At first, I ran validate.sh on my machine and the result was very close to 0.8 so it didn't pass the check up here (on Github)
Then I refactored the function, also referring to other classmates' implementations; however, I passed all the four tests on my own machine (ratio16/ratio0 is about 0.5-0.6 which is smaller than 0.8) but still failed the check up here.

Could you please let me know if there is still anything I can improve? or what I have done wrong?
Thank you.

Reply:
It's OK for the timing test to fail on github action whose machines are noisy.  If the runtime works ok on the machines curated by the grader, I think that's good enough.  Please double check with the grader.